### PR TITLE
#18424 : Content Types from very old customer datasets present lots o…

### PR DESCRIPTION
**Pull request by jcastro-dotcms**
…f data arrangements and values that are no longer valid. This is causing the reindex to fail and get stuck. The code changes in this PR are the following:

1. The `ContentletIndexAPIImpl` correctly catches any error during the indexation of a contentlet. Therefore, a failing contentlet will be flagged as failed now, instead of keeping the reindex running endlessly.
2. More useful logging in the `ImplClassFieldTransformer` class.
3. Handling NPE situations where the `index_val` column in the `dist_reindex_journal` table might contain null values.
4. Fixing code to declare failed records as the ones whose `priority` is GREATER than 300, not GREATER OR EQUAL.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java [hunk 1]
```diff
@@ -13,7 +15,7 @@
 
   final List<Field> genericFields;
 
-  public ImplClassFieldTransformer(Field f) {
+  public ImplClassFieldTransformer(final Field f) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java:18]: Codacy Issue found: [Avoid variables with short names like f](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java [hunk 2]
```diff
@@ -758,37 +758,43 @@ public BulkRequest appendBulkRequest(BulkRequest bulkRequest, final ReindexEntry
...
+                Logger.debug(this, String.format("Indexing id: '%s', priority: '%s'", contentlet.getInode(), idx
+                        .getPriority()));
+                contentlet.setIndexPolicy(IndexPolicy.DEFER);
                 addBulkRequest(bulk, ImmutableList.of(contentlet), idx.isReindex());
-
-            } catch (Exception e) {
-                APILocator.getReindexQueueAPI().markAsFailed(idx, e.getMessage());
-
             }
+        } catch (final Exception e) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java:795]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java [hunk 3]
```diff
@@ -23,28 +25,39 @@ public ImplClassFieldTransformer(final List<Field> fields) {
...
 
   private static Field impleClass(final Field genericField) {
-    FieldBuilder builder = null;
+    FieldBuilder builder;
     try {
       builder = FieldBuilder.builder(genericField);
       return builder.build();
-    } catch (Exception e) {
-      throw new DotStateException(e.getMessage(), e);
+    } catch (final Exception e) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java:41]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

**Review by nollymar — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

**Review by freddyucv — COMMENTED** [review]
any test?

**Review by freddyucv — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java:795]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java:41]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java:18]: Codacy Issue found: [Avoid variables with short names like f](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.Map;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class ReindexQueueAPITest {
+
+    private static ReindexQueueAPI reindexQueueAPI;
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java:25]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotmarketing/common/reindex/ReindexQueueFactory.java [hunk 5]
```diff
@@ -135,24 +139,36 @@ protected void deleteFailedRecords() throws DotDataException {
...
+        dc.setSQL("SELECT id, ident_to_index, priority, index_val, time_entered FROM dist_reindex_journal WHERE priority > ?");
         dc.addParam(ReindexQueueFactory.Priority.REINDEX.dbValue());
-        List<Map<String, Object>> failedRecords = dc.loadObjectResults();
-        List<ReindexEntry> failed = new ArrayList<>();
-        for (Map<String, Object> map : failedRecords) {
-            ReindexEntry ridx = new ReindexEntry()
-                    .setId(Long.parseLong((String) map.get("id")))
+        final List<Map<String, Object>> failedRecords = dc.loadObjectResults();
+        final List<ReindexEntry> failed = new ArrayList<>();
+        long id;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/common/reindex/ReindexQueueFactory.java:147]: Codacy Issue found: [Avoid variables with short names like id](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotmarketing/common/reindex/ReindexQueueFactory.java [hunk 6]
```diff
@@ -135,24 +139,36 @@ protected void deleteFailedRecords() throws DotDataException {
...
+        int priority;
+        for (final Map<String, Object> map : failedRecords) {
+            final String indexVal = UtilMethods.isSet(map.get("index_val")) ? String.class.cast(map.get("index_val"))
+                    : StringUtils.EMPTY;
+
+            if (DbConnectionFactory.isOracle()) {
+                BigDecimal rowVal = (BigDecimal) map.get("id");
+                id = new Long(rowVal.toPlainString());
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/common/reindex/ReindexQueueFactory.java:155]: Codacy Issue found: [Avoid instantiating Long objects.Call Long.valueOf() instead](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    }
+
+    /**
+     * Gets a {@link ReindexEntry} given its id
+     * @param recordId
+     * @return
+     * @throws DotDataException
+     */
+    private ReindexEntry getReindexEntry(String recordId) throws DotDataException {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java:111]: Codacy Issue found: [Parameter 'recordId' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotmarketing/common/reindex/ReindexQueueFactory.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/common/reindex/ReindexQueueFactory.java:155]: Codacy Issue found: [Avoid instantiating Long objects.Call Long.valueOf() instead](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java:795]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java:111]: Codacy Issue found: [Parameter 'recordId' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java:18]: Codacy Issue found: [Avoid variables with short names like f](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java:25]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotmarketing/common/reindex/ReindexQueueFactory.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/common/reindex/ReindexQueueFactory.java:147]: Codacy Issue found: [Avoid variables with short names like id](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java:41]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java:795]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java:25]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

File: dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/transform/field/ImplClassFieldTransformer.java:41]: Codacy Issue found: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5458294)

**Pull request merged**